### PR TITLE
Feature/predictoor iteration 2

### PR DESCRIPTION
**Pull request by alexcos20**
Fixes # .

Changes proposed in this PR:

- add floatValue  to submitTrueVal, for better accuracy tracking
- allow owner to cancel an epoch by passing cancelRound = true in submitTrueVal
- make sure that feeCollector gets fixed rate fees, slashed stakes, revenue per epoch if no predictoors  (and not ERC20Deployer)
- handle edge case when all stakers are wrong and will be slashed  (feeCollector gets all stakes)

Tests:

- [x] Write test for edge case
- [x] Write test for "allow owner to cancel an epoch by passing cancelRound"

### CODE REVIEW

**Review by trentmc — APPROVED** [review]

File: contracts/templates/ERC20Template3.sol [hunk 1]
```diff
@@ -110,9 +111,13 @@ contract ERC20Template3 is
     mapping(uint256 => uint256) private aggregatedPredictedValuesNumer;
     mapping(uint256 => uint256) private aggregatedPredictedValuesDenom;
     mapping(uint256 => bool) public trueValues; // true values submited by owner
+    mapping(uint256 => uint256) public floatValues; // real values submited by owner
     mapping(uint256 => Status) public epochStatus; // status of each epoch
     mapping(uint256 => uint256) private subscriptionRevenueAtBlock; //income registred
     mapping(address => Subscription) public subscriptions; // valid subscription per user
+    mapping(uint256 => uint256) private noOfTruePredictionsPerEpoch; // how many Ture predictions we have per epoch
```
[the remark this answers is not in the document]
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol:118]: commit on the way

[the remark this answers is not in the document]
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol]: see https://github.com/oceanprotocol/contracts/pull/755/commits/d4804c74658bb9fb0e82a72649a57eb2b793797a

**Comment by alexcos20** [comment]
> ```solidity
> roundSumStakes 
> ```

done

File: contracts/templates/ERC20Template3.sol [hunk 2]
```diff
@@ -1067,34 +1072,70 @@ contract ERC20Template3 is
...
+     * @param cancelRound If true, cancel that epoch
+     */
     function submitTrueVal(
         uint256 blocknum,
-        bool trueValue
+        bool trueValue,
+        uint256 floatValue,
+        bool cancelRound
     ) external onlyERC20Deployer {
         // TODO, is onlyERC20Deployer the right modifier?
```
[the remark this answers is not in the document]
Comment (alexcos20) [reply contracts/templates/ERC20Template3.sol]: done

**Comment by alexcos20** [comment]
> > > I'm having trouble understanding the purpose of `floatValues` as it doesn't seem to be utilized in the code.
> > 
> > 
> > ```
> > emit TruevalSubmitted(slot, trueValue,floatValue,epochStatus[slot]);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > at https://github.com/oceanprotocol/contracts/blob/feature/predictoor_iteration_2/contracts/templates/ERC20Template3.sol#LL1138C9-L1138C77
> 
> The mapping looks redundant. It's only used at:
> 
> ```solidity
> floatValues[slot] = floatValue;
> ```
> 
> Considering it's not used anywhere else and we can query the events to get `floatValue` I suggest we remove the mapping. What do you think?

yeap, removed
